src/wp-includes/kses.php

```diff
@@ -1674,18 +1674,41 @@ function wp_kses_check_attr_val( $value, $vless, $checkname, $checkvalue ) {
 /**
  * Sanitizes a string and removed disallowed URL protocols.
  *
- * This function removes all non-allowed protocols from the beginning of the
+ * This function first tries to return early by checking for a standard http(s)
+ * url, and otherwise removes all non-allowed protocols from the beginning of the
  * string. It ignores whitespace and the case of the letters, and it does
  * understand HTML entities. It does its work recursively, so it won't be
  * fooled by a string like `javascript:javascript:alert(57)`.
  *
+ * The regular expression is based on the pattern from @diegoperini compared
+ * here: https://mathiasbynens.be/demo/url-regex
+ *
  * @since 1.0.0
  *
  * @param string   $string            Content to filter bad protocols from.
  * @param string[] $allowed_protocols Array of allowed URL protocols.
  * @return string Filtered content.
  */
 function wp_kses_bad_protocol( $string, $allowed_protocols ) {
+	// Detect standard HTTP(S) URL and return early.
+	$regex = '_^(?:(?<protocol>https?)://)(?:\S+(?::\S*)?@)?(?:(?!10(?:\.\d{1,3}){3})(?!127(?:\.\d{1,3}){3})(?!169\.254(?:\.\d{1,3}){2})(?!192\.168(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\x{00a1}-\x{ffff}0-9]+-?)*[a-z\x{00a1}-\x{ffff}0-9]+)(?:\.(?:[a-z\x{00a1}-\x{ffff}0-9]+-?)*[a-z\x{00a1}-\x{ffff}0-9]+)*(?:\.(?:[a-z\x{00a1}-\x{ffff}]{2,})))(?::\d{2,5})?(?:/[^\s]*)?$_iuS';
```

Comment:
This string should preferably be a constant. But if there's no clear place to turn it into a constant, it could at the very least be made static here. This ensures the entire memory allocation and string construction only happens once.
```suggestion
	static $regex = '_^(?:(?<protocol>https?)://)(?:\S+(?::\S*)?@)?(?:(?!10(?:\.\d{1,3}){3})(?!127(?:\.\d{1,3}){3})(?!169\.254(?:\.\d{1,3}){2})(?!192\.168(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\x{00a1}-\x{ffff}0-9]+-?)*[a-z\x{00a1}-\x{ffff}0-9]+)(?:\.(?:[a-z\x{00a1}-\x{ffff}0-9]+-?)*[a-z\x{00a1}-\x{ffff}0-9]+)*(?:\.(?:[a-z\x{00a1}-\x{ffff}]{2,})))(?::\d{2,5})?(?:/[^\s]*)?$_iuS';
```

Reply:
Thanks @schlessera for the review. Changes make sense to me.